### ISSUE TITLE
IntlProvider creates intl object multiple times, messing up with shallow equality and sCU

### ISSUE THREAD

**Issue by leoasis** [formatjs/react-intl#818]
Here's a jsfiddle to repro the issue: https://jsfiddle.net/nhnuobwt/3/

Check that the output shows that the `Comp` component rendered 4 times, even though the object has a shouldComponentUpdate hook shallow checking for the `intl` prop not changing.
To see the problem, go to the `ReactIntl.IntlProvider` call and comment it, and uncomment the `FakeIntlProvider` line. Run it again and you'll see 4 times in the output again, because I tried to mimic the issue in that component. As a last step, go to the `FakeIntlProvider`'s `getChildContext` implementation, and comment the `intl` prop object literal, and uncomment the one that gets its value from the state. Now run it again, and you'll see the expected output: the `Comp` component is only rendered once.

Basically the problem is that the `intl` object received in components gets recreated every time the `IntlProvider` is rendered, because the `getChildContext` method is being called again by React to build the context that contains the `intl` object.

This can be solved by creating and holding the intl object in the `IntlProvider` state, or at least memoizing the method so that it returns the same `intl` instance given the same props and state.

Hope this is somewhat clear, if not let me know if I should clarify or explain a bit more.

**Comment by ericf** [comment]
I'm curious to understand why there's a concern about `<IntlProvider>` being rendered many times. How is your app structured such that this happens? Usually it's towards the root of the tree and doesn't end up being re-rendered. Also it implements sCU, and should only be re-rendering when its props change.

**Comment by leoasis** [comment]
The problem is not that the `IntlProvider` is re-rendered, but the problem that causes: it makes the `intl` object in context be recreated, causing shallow comparison in inner components that get the prop injected to fail.

I'm not entirely sure what are the semantics for when `getChildContext` is called by React, but it appears to be called again when the `IntlProvider` is sent new props. We may need to investigate a bit more to understand exactly when it is that React calls `getChildContext` (or maybe you already know). I assume it happens at least whenever the component is re-rendered.

In this case though, the `IntlProvider` will re-render every time since it will return false in its sCU since the children prop is always different. 

I understand what you say with respect to `IntlProvider` usually appearing at the top of the component tree, but in our app this is currently inside an `App` root component that gets re-rendered for example whenever a route changes (to change the page component to render). This causes the `IntlProvider` to re-render, even though the props passed are the same. But that only thing causes the `intl` object in context to be recreated, causing these issues.

I think this behavior is not what you'd expect, and if the `IntlProvider` needs to be at the top and never to be re-rendered, then it should be explained in the docs, but again, this doesn't make sense since the lib also supports nested providers, so those ones at least will be re-rendered at some point. And also, the fix is simple enough: memoize the construction of the intl object.

In any case I think this should be fixed either with docs, adding a warning that you shouldn't be re-rendering the provider, or just fixing it inside the provider itself.

What do you think?

**Comment by ericf** [comment]
> This causes the IntlProvider to re-render, even though the props passed are the same.

This is what doesn't make sense to me since `<IntlProvider>` implements `shouldComponentUpdate()`.

> And also, the fix is simple enough: memoize the construction of the intl object.

That's effectively what it already does by implementing sCU.

Are you sure you're always passing the same props to `<IntlProvider>`, maybe you're recreating your `messages` object each time your root `<App>` component re-renders?

**Comment by leoasis** [comment]
> This is what doesn't make sense to me since <IntlProvider> implements shouldComponentUpdate().

Yes, but the `children` is also a prop and it is always created in each render since it is a react element.

> That's effectively what it already does by implementing sCU.

Not quite, since the `intl` object creation only depends on some state and props, not all of them (e.g. it doesn't depend on the children)

Check the example I provided in the PR description where this is happening.

**Comment by ericf** [comment]
@leoasis thanks for providing this additional info. I'm working on some related changes which touch code in `<IntlProvider>`, so I'll look at addressing what you're seeing at the same time.

**Comment by leoasis** [comment]
@ericf awesome thanks! Let me know if you need help as well, if you prefer I can send a PR to fix this issue.

**Comment by longlho** [comment]
closing due to stale. Feel free to reopen if this issue arises in 3.0.0-beta.8
